### PR TITLE
Removes latejoin slots and merges them with the normal job slots if they would be given

### PR DESCRIPTION
**Pull request by DGamerL**
## What Does This PR Do
Makes it so if there would be highpop slots toggled, they are added to the total job slots before jobs are handed out, instead of after the round has started.

## Why It's Good For The Game
Fun fact: the best way to roll most jobs during highpop is to wait out roundstart and then quickly joing game for the job you want. Not very good design.

## Testing
It compiled, if anyone knows how to test it properly I'd be glad to hear it

## Changelog
:cl:
fix: Fixed highpop slots opening after the round has started
/:cl:

### CODE REVIEW

**Comment by matttheficus** [comment]
Thank god. A relic of lowpop being the standard.

**Review by Henri215 — APPROVED** [review]

File: code/controllers/subsystem/SSticker.dm [hunk 1]
```diff
@@ -226,7 +226,19 @@ SUBSYSTEM_DEF(ticker)
 	mode.pre_pre_setup()
 	var/can_continue = FALSE
 	can_continue = mode.pre_setup() //Setup special modes. This also does the antag fishing checks.
+
+	// Now that every other piece of the round has initialized, lets setup player job scaling // Not completely true anymore, but if we do this later we only get the latejoin slots after the round has started
```
Comment (warriorstar-orion) [inline code/controllers/subsystem/SSticker.dm:230]: There's no need to have a comment commenting about how another comment is wrong. 
```suggestion
	// Enable highpop slots just before we distribute jobs.
```

**Comment by warriorstar-orion** [comment]
> It compiled, if anyone knows how to test it properly I'd be glad to hear it

@DGamerL  Blackbox records the `manifest` data, which is a nested tally containing job assignments and whether they were roundstart or latejoin:

https://github.com/ParadiseSS13/Paradise/blob/cdef637e7f018d439d704eb81a3d603171248e10/code/controllers/subsystem/SSjobs.dm#L453

Using that and the `testmerged_prs` data, we can TM this and see the before-and-after of how many jobs were assigned roundstart versus latejoin when this PR is testmerged. We should expect to consistently see that rounds with this TM has a greater number of roundstart jobs assigned. I know that's not "testing" in the way we expect before a PR is submitted but we're pretty limited when it comes to testing or faking things that require large numbers of clients. You could attempt to make a fake client/mind that you add a bunch to just before roundstart but that sounds like a pain in the ass and not worth the effort.

**Comment by DGamerL** [comment]
> > It compiled, if anyone knows how to test it properly I'd be glad to hear it
> 
> @DGamerL Blackbox records the `manifest` data, which is a nested tally containing job assignments and whether they were roundstart or latejoin:
> 
> https://github.com/ParadiseSS13/Paradise/blob/cdef637e7f018d439d704eb81a3d603171248e10/code/controllers/subsystem/SSjobs.dm#L453
> 
> Using that and the `testmerged_prs` data, we can TM this and see the before-and-after of how many jobs were assigned roundstart versus latejoin when this PR is testmerged. We should expect to consistently see that rounds with this TM has a greater number of roundstart jobs assigned. I know that's not "testing" in the way we expect before a PR is submitted but we're pretty limited when it comes to testing or faking things that require large numbers of clients. You could attempt to make a fake client/mind that you add a bunch to just before roundstart but that sounds like a pain in the ass and not worth the effort.

Oh that's interesting, didn't know we had that. Should I ask Steelslayer/another headcoder to have this PR be TM'ed a bit (maybe slightly ahead of the queue) for prod testing?

**Comment by warriorstar-orion** [comment]
> > > It compiled, if anyone knows how to test it properly I'd be glad to hear it
> > 
> > 
> > @DGamerL Blackbox records the `manifest` data, which is a nested tally containing job assignments and whether they were roundstart or latejoin:
> > https://github.com/ParadiseSS13/Paradise/blob/cdef637e7f018d439d704eb81a3d603171248e10/code/controllers/subsystem/SSjobs.dm#L453
> > 
> > Using that and the `testmerged_prs` data, we can TM this and see the before-and-after of how many jobs were assigned roundstart versus latejoin when this PR is testmerged. We should expect to consistently see that rounds with this TM has a greater number of roundstart jobs assigned. I know that's not "testing" in the way we expect before a PR is submitted but we're pretty limited when it comes to testing or faking things that require large numbers of clients. You could attempt to make a fake client/mind that you add a bunch to just before roundstart but that sounds like a pain in the ass and not worth the effort.
> 
> Oh that's interesting, didn't know we had that. Should I ask Steelslayer/another headcoder to have this PR be TM'ed a bit (maybe slightly ahead of the queue) for prod testing?

Yeah, request a TM, and I can help munging the data for the relevant rounds. Dunno if you'd get to skip the queue though, though this is a pretty important change if it works as expected.

File: code/controllers/subsystem/SSticker.dm [hunk 2]
```diff
@@ -226,7 +226,19 @@ SUBSYSTEM_DEF(ticker)
...
+	var/highpop_trigger = 80
+
+	if(playercount >= highpop_trigger)
+		log_debug("Playercount: [playercount] versus trigger: [highpop_trigger] - loading highpop job config")
+		SSjobs.LoadJobs(TRUE)
+	else
+		log_debug("Playercount: [playercount] versus trigger: [highpop_trigger] - keeping standard job config")
+
 	SSjobs.DivideOccupations() //Distribute jobs
+
```
Comment (farie82) [inline code/controllers/subsystem/SSticker.dm:241]: I'd move this down to below the `can_continue` check. Do your checks first, then actually do the setup

**Review by lewcc — APPROVED** [review]
lgtm

**Comment by warriorstar-orion** [comment]
@DGamerL I'm not quite sure this is working as expected. Rounds 37042, 37045, and 37046 were all rounds when two players had janitor set to high in their job preferences but only one janitor joined roundstart.

Things I also checked:

- The rounds all were autotot and cult, so there were no job restrictions or assignments of non-crew antags.
- The rounds all took place on Cyberiad, where there are two janitor spawn points.
- The players were not jobbanned (their high preference wouldn't count otherwise)

**Comment by DGamerL** [comment]
> @DGamerL I'm not quite sure this is working as expected. Rounds 37042, 37045, and 37046 were all rounds when two players had janitor set to high in their job preferences but only one janitor joined roundstart.
> 
> Things I also checked:
> 
> * The rounds all were autotot and cult, so there were no job restrictions or assignments of non-crew antags.
> * The rounds all took place on Cyberiad, where there are two janitor spawn points.
> * The players were not jobbanned (their high preference wouldn't count otherwise)
> 
> EDIT:
> 
> * I also confirmed with affected that the settings in the example config.toml are what's on prod
> * It's possible these people got assigned to a command job, I think? But the fact it happened in three rounds already with the same role makes me think otherwise.

Hmmm weird, got any idea where the issue might be then? I was mostly going off of what AA said in an old forum post, so I haven't looked much farther into it than what is edited here

File: code/controllers/subsystem/SSjobs.dm [hunk 3]
```diff
@@ -275,7 +275,8 @@ SUBSYSTEM_DEF(jobs)
 	var/watch = start_watch()
 	//Setup new player list and get the jobs list
 	Debug("Running DO")
-	SetupOccupations()
+	if(!occupations.len)
```
Comment (SteelSlayer) [inline code/controllers/subsystem/SSjobs.dm:278]:
```suggestion
	if(!length(occupations))
```

Comment (DGamerL) [reply code/controllers/subsystem/SSjobs.dm:278]: How did I miss this...

**Comment by warriorstar-orion** [comment]
Thanks @HMBGERDO! It looks like we're good now

2023_08_26__18_02_06__Untitled2

Two roundstart janitors on round 37092.

**Review by S34NW — APPROVED** [review]
Aight this should be fine now, lets go to mars

**Pull request merged**